### ISSUE TITLE
secio: WOAH: same keys (AND same nonce: 1/(2^128) chance!)

### ISSUE THREAD

**Issue by torarnv** [ipfs/go-ipfs#1016]
Running `ipfs daemon` on OSX 10.10 logs a scary error from [p2p/crypto/secio/protocol.go](https://github.com/ipfs/go-ipfs/blob/79360bbd32d8a0b9c5ab633b5a0461d9acd0f477/p2p/crypto/secio/protocol.go#L250) after about a minute:

```
Initializing daemon...
API server listening on /ip4/127.0.0.1/tcp/5001
Gateway (readonly) server listening on /ip4/127.0.0.1/tcp/8080
18:17:26.861 ERROR      secio: WOAH: same keys (AND same nonce: 1/(2^128) chance!). <autogenerated>:24
^CReceived interrupt signal, shutting down...
18:17:30.546 ERROR     ipnsfs: leveldb: closed <autogenerated>:24
```

Reproducible if I restart the daemon.

**Comment by torarnv** [comment]
Can provide `events.log` if that helps and it doesn't contain anything private.

**Comment by whyrusleeping** [comment]
I'm pretty sure we have only ever seen this on OSX.. Very strange.

**Comment by torarnv** [comment]
Unfortunately I can't reproduce anymore by just restarting daemon, will try to run with --debug to get more info if it happens again.

**Comment by whyrusleeping** [comment]
@torarnv dont worry about providing events, we've been seeing this issue for quite some time now, and my personal hunch is that its caused by a bad RNG. cant prove it though...

**Comment by whyrusleeping** [comment]
either that, or the node somehow managed to dial itself

**Comment by whyrusleeping** [comment]
Actually, the more i look at this code, the more i think this is just a peer dialing itself.

**Comment by jbenet** [comment]
@whyrusleeping it definitely is from connecting to self-- as the keys are the same. But the nonces SHOULD still be different. one is generated per each side of the connection (even if you're at both sides). 

it's not sync problems in rand, i looked, it mutexes correctly.

it's definitely our bug-- even with a broken RNG wouldn't get two nonces that large being _exactly the same_.

So i think what's happening here is that "both" sides are in fact one side. it's not only a node connecting to self (a Dial landing in a Listen), the `io.ReadWriter` is basically a loopback/pipe. This is corroborated by the fact that this `log.Error` only appears once per time.

Next steps:
- [ ] find a reliable way to repro
- [ ] insert this as a debug line:

```go
log.Error("WOAH: same keys (AND same nonce: 1/(2^128) chance!).")
log.Error("k1: %v, k2: %v, insecure: %v, insecureM %v", k1, k2, insecure, insecureM)
```

**Comment by torarnv** [comment]
> This is corroborated by the fact that this `log.Error` only appears once per time.

The first time I launched `ipfs daemon` i saw the error twice. Once shortly after startup, and then again some 10-15 minutes later.

I've added the extra debug logging and will try to catch the bug again.

**Comment by jbenet** [comment]
@torarnv thanks! if you PR, we can all run the extra debugging

**Comment by whyrusleeping** [comment]
@jbenet the *same* handshake call is handling both sides of the connection i think.

**Comment by whyrusleeping** [comment]
otherwise, we would see the message twice in rapid succession.

**Comment by jbenet** [comment]
Right exactly. 



—

**Comment by whyrusleeping** [comment]
I think that if we see same key and nonce, we can assume that its a peer connecting to itself, but what do we want to do in that case?

**Comment by jbenet** [comment]
We should probably silence the message as it's wrong (bump down to debug).

I think this has to do with two problems.

1. Peers connect to self. I think this comes from dialing local addresses of other peers, that happen to clash. (e.g. `/ip4/192.168.0.100/tcp/4001`). Without randomizing the swarm port, we'll see a lot of this happen naturally. It's a waste because of the dials, but not a huge one. We could try to optimize by having the Swarm discern when it will be connecting to self and prevent that. this has to be done carefully though because preventing connection to self entirely is one of those silly "lack of foresight" things. (e.g. NATs that not only don't implement hairpinning, but actively _drop_ those packets. i have one)

2. There may be something else going on here. The secio handshake is happening on the _same function_ which raises questions about when that happens. (note this is different from a dialing self, and hitting the listener, as _i think_ that will cause two different calls to secio (and thus would have different nonces).

**Comment by whyrusleeping** [comment]
with reuseport a socket can connect to itself.

**Comment by jbenet** [comment]
> with reuseport a socket can connect to itself.

as usual, you're exactly right. it's a dial out, but it's not even hitting the listener.

**Comment by jbenet** [comment]
Fixed in #1073
